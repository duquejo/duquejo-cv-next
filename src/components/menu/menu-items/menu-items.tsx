'use client';

import {
  SidebarMenu,
  SidebarMenuButton,
  SidebarMenuItem,
  useSidebar,
} from '@/components/ui/sidebar';
import type { MenuItem } from '@/interfaces';
import { Link, usePathname } from '@/i18n/routing';

interface Props {
  items: MenuItem[];
  size?: 'default' | 'sm' | 'lg';
}

export const MenuItems = ({ items, size = 'default' }: Props) => {
  const pathName = usePathname();
  const { setOpenMobile } = useSidebar();

  return (
    <SidebarMenu>
      {items.map((item) => (
        <SidebarMenuItem key={item.title}>
          <SidebarMenuButton
            asChild
            isActive={pathName === item.url}
            size={size}
            className="focus-visible:ring-primary focus-visible:ring-1"
          >
            <Link
              href={item.url as Parameters<typeof Link>['0']['href']}
              onClick={() => setOpenMobile(false)}
            >
              <span>{item.title}</span>
            </Link>
          </SidebarMenuButton>
        </SidebarMenuItem>
      ))}
    </SidebarMenu>
  );
};
