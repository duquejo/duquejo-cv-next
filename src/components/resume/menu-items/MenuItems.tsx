'use client';

import {
  SidebarMenu,
  SidebarMenuButton,
  SidebarMenuItem,
  useSidebar,
} from '@/components/ui/sidebar';
import Link from 'next/link';
import { usePathname } from 'next/navigation';
import type { MenuItem } from '@/interfaces';

interface Props {
  items: MenuItem[];
  size?: 'default' | 'sm' | 'lg';
}

export const MenuItems = ({ items, size = 'default' }: Props) => {
  const pathName = usePathname();
  const { setOpenMobile } = useSidebar();

  return (
    <SidebarMenu>
      {items.map((item) => (
        <SidebarMenuItem key={item.title}>
          <SidebarMenuButton asChild isActive={pathName === item.url} size={size}>
            <Link href={item.url} onClick={() => setOpenMobile(false)}>
              {item.icon}
              <span>{item.title}</span>
            </Link>
          </SidebarMenuButton>
        </SidebarMenuItem>
      ))}
    </SidebarMenu>
  );
};
