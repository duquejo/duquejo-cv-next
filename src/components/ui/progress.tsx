'use client';

import * as React from 'react';
import { useEffect, useState } from 'react';
import * as ProgressPrimitive from '@radix-ui/react-progress';

import { cn } from '@/lib/utils';

const Progress = React.forwardRef<
  React.ComponentRef<typeof ProgressPrimitive.Root>,
  React.ComponentPropsWithoutRef<typeof ProgressPrimitive.Root>
>(({ className, value = 0, color = '#000', ...props }, ref) => {
  const [progress, setProgress] = useState(0);

  useEffect(() => {
    const intervalSpeed = 30;
    const incrementSpeed = 1;

    const progressInterval = setInterval(() => {
      setProgress((prevProgress) => {
        const updatedProgress = prevProgress + incrementSpeed;
        if (updatedProgress >= value!) {
          clearInterval(progressInterval);
          return value!;
        }
        return updatedProgress;
      });
    }, intervalSpeed);

    return () => clearInterval(progressInterval);
  }, [value]);

  return (
    <ProgressPrimitive.Root
      ref={ref}
      className={cn('relative h-2 w-full overflow-hidden rounded-full bg-secondary', className)}
      {...props}
    >
      <ProgressPrimitive.Indicator
        className="h-full w-full flex-1 transition-all"
        style={{ backgroundColor: color, transform: `translateX(-${100 - (progress || 0)}%)` }}
      />
    </ProgressPrimitive.Root>
  );
});
Progress.displayName = ProgressPrimitive.Root.displayName;

export { Progress };
