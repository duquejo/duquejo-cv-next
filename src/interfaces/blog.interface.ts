import { ReactNode } from 'react';

export interface BlogPost {
  slug: string;
  title: string;
  excerpt: string;
  category: string;
  tags: string[];
  publishDate: string;
  readingTime: string;
  coverImage: string;
}

export interface BlogPostResult {
  Post: () => ReactNode;
  metadata: BlogPost;
}
