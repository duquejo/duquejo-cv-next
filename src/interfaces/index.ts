export * from './experience.interface';
export * from './menu-items.interface';
export * from './main-languages.interface';
export * from './services.interface';
export * from './knowledge.interface';
