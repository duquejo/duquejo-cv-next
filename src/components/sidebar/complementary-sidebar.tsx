import { ThemeToggle } from '@/components/theme/theme-toggle';
import { SidebarTrigger } from '@/components/ui/sidebar';
import { SheetTrigger } from '@/components/ui/sheet';
import { Lightbulb } from 'lucide-react';
import { Button } from '@/components/ui/button';
import * as React from 'react';
import { PdfGeneratorToggle } from '@/components/pdf/pdf-toggle';
import { useTranslations } from 'next-intl';

interface Props {
  outerClassName?: string;
  innerClassName?: string;
}

export const ComplementarySidebar = ({ outerClassName, innerClassName }: Props) => {
  const t = useTranslations('Sidebar.complementary');

  return (
    <div role="complementary" className={outerClassName}>
      <div className={innerClassName}>
        <SidebarTrigger
          variant="outline"
          className="h-8 w-8 lg:h-10 lg:w-10 hidden sm:flex"
          title={t('sidebar.title')}
        />
        <PdfGeneratorToggle
          variant="outline"
          className="h-8 w-8 lg:h-10 lg:w-10"
          buttonTitle={t('pdf.title')}
          formTitle={t('pdf.dialog.title')}
          formDescription={t('pdf.dialog.description')}
          formButtonText={t('pdf.dialog.button')}
          formButtonTextLoading={t('pdf.dialog.loading')}
        />
        <ThemeToggle
          variant="outline"
          className="h-8 w-8 lg:h-10 lg:w-10"
          title={t('theme.title')}
          light={t('theme.light')}
          dark={t('theme.dark')}
          system={t('theme.system')}
        />
        <SheetTrigger asChild>
          <Button
            variant="outline"
            className=" h-8 w-8 lg:h-10 lg:w-10 hidden sm:flex"
            title={t('activity.title')}
          >
            <Lightbulb />
            <span className="sr-only">{t('activity.title')}</span>
          </Button>
        </SheetTrigger>
      </div>
    </div>
  );
};
