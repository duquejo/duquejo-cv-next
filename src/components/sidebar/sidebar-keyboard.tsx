import { cn } from '@/lib';
import { Kbd, KbdGroup } from '@/components/ui/kbd';

export const SidebarKeyboard = ({ className }: React.ComponentProps<'div'>) => {
  return (
    <div className={cn('flex-col items-center gap-4', className)}>
      <p className="text-xs text-muted-foreground">
        Use{' '}
        <KbdGroup>
          <Kbd>Ctrl</Kbd>
          <span>+</span>
          <Kbd>B</Kbd>
        </KbdGroup>{' '}
        to toggle the sidebar
      </p>
    </div>
  );
};
