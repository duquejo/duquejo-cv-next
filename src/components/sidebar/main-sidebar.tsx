import {
  Sidebar,
  SidebarContent,
  SidebarFooter,
  SidebarGroup,
  SidebarGroupContent,
  SidebarGroupLabel,
  SidebarHeader,
  SidebarMenu,
  SidebarMenuItem,
  SidebarSeparator,
} from '@/components/ui/sidebar';
import { MenuAvatar } from '@/components/menu/menu-avatar/menu-avatar';
import { useTranslations } from 'next-intl';
import { MenuItems } from '@/components/menu/menu-items/menu-items';
import { ComplementarySidebar } from '@/components/sidebar/complementary-sidebar';
import { SocialItemsHorizontal } from '@/components/social/social-items-horizontal';

export const MainSidebar = () => {
  const t = useTranslations('Sidebar');

  return (
    <Sidebar collapsible="icon" variant="floating" role="navigation" tabIndex={0}>
      <SidebarHeader>
        <SidebarGroup>
          <SidebarMenu className="gap-y-3">
            <SidebarMenuItem className="flex flex-col place-items-center gap-y-2">
              <MenuAvatar role={t('role')} />
            </SidebarMenuItem>
          </SidebarMenu>
        </SidebarGroup>
        <SidebarSeparator />
        <SidebarGroup className="p-0">
          <SidebarGroupLabel className="font-semibold tracking-wider text-foreground">
            {t('contact.title')}
          </SidebarGroupLabel>
          <SidebarGroupContent>
            <SocialItemsHorizontal />
          </SidebarGroupContent>
        </SidebarGroup>
        <SidebarSeparator />
      </SidebarHeader>
      <SidebarContent className="justify-between">
        <SidebarGroup>
          <SidebarGroupLabel className="font-semibold tracking-wider text-foreground">
            {t('professional.title')}
          </SidebarGroupLabel>
          <SidebarGroupContent>
            <MenuItems items={t.raw('professional.links')} />
          </SidebarGroupContent>
        </SidebarGroup>
      </SidebarContent>
      <SidebarFooter>
        <SidebarGroup className="sm:hidden">
          <SidebarSeparator className="mx-0" />
          <SidebarGroupLabel className="px-0 font-semibold tracking-wider text-foreground">
            {t('footer.mobile.title')}
          </SidebarGroupLabel>
          <SidebarGroupContent>
            <SidebarMenu>
              <SidebarMenuItem>
                <ComplementarySidebar innerClassName="flex flex-row gap-x-2" />
              </SidebarMenuItem>
            </SidebarMenu>
          </SidebarGroupContent>
        </SidebarGroup>
      </SidebarFooter>
    </Sidebar>
  );
};
