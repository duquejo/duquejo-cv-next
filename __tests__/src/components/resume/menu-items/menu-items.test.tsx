import { MenuItems } from '@/components/resume/menu-items/MenuItems';
import { fireEvent, render, screen } from '@testing-library/react';
import { MenuItem } from '@/interfaces';
import { SidebarProvider, useSidebar } from '@/components/ui/sidebar';

// Partial mocking
vi.mock('@/components/ui/sidebar', async (importOriginal) => ({
  ...(await importOriginal()),
  useSidebar: vi.fn(),
}));

describe('<MenuItems /> tests', () => {
  const setOpenMobileMock = vi.fn();

  const mockItems: MenuItem[] = [
    {
      title: 'Foo',
      url: 'foo.bar',
      icon: <svg />,
    },
    {
      title: 'Bar',
      url: 'bar.baz',
      icon: <svg />,
    },
  ];

  beforeEach(() => {
    vi.clearAllMocks();

    Object.defineProperty(window, 'location', {
      writable: true,
      value: {
        href: 'mock',
      },
    });

    window.matchMedia = vi.fn().mockImplementation((query) => ({
      matches: query === '(min-width: 768px)',
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    }));

    vi.mocked(useSidebar).mockReturnValue({
      state: 'expanded',
      open: true,
      isMobile: true,
      openMobile: true,
      setOpen: vi.fn(),
      toggleSidebar: vi.fn(),
      setOpenMobile: setOpenMobileMock,
    });
  });

  it('Should match the snapshot with the default args', () => {
    const { container } = render(
      <SidebarProvider>
        <MenuItems items={mockItems} />
      </SidebarProvider>,
    );
    expect(container).toMatchSnapshot();

    mockItems.forEach((item) => {
      const aTag = screen.getByRole('link', { name: item.title });
      expect(aTag).toHaveAttribute('href', item.url);
      expect(aTag).toContainHTML('svg');
    });
  });

  it('Should match the snapshot with the available args', () => {
    const { container } = render(
      <SidebarProvider>
        <MenuItems items={mockItems} size="lg" />
      </SidebarProvider>,
    );
    expect(container).toMatchSnapshot();

    mockItems.forEach((item) => {
      expect(screen.getByRole('link', { name: item.title })).toHaveAttribute('data-size', 'lg');
    });
  });

  it('Should handle the click events', () => {
    render(
      <SidebarProvider>
        <MenuItems items={mockItems} />
      </SidebarProvider>,
    );

    const item = screen.getByRole('link', { name: mockItems[0].title });

    /**
     * @see Error: Not implemented: navigation (except hash changes)
     * Adverse routing while clicking & adding the setOpenMobile action
     */
    item.addEventListener('click', (e) => {
      e.preventDefault();
    });

    fireEvent.click(item);

    expect(setOpenMobileMock).toHaveBeenCalledWith(false);
  });
});
