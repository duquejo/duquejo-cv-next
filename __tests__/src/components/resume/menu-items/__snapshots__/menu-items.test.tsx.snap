// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<MenuItems /> tests > Should match the snapshot with the available args 1`] = `
<div>
  <div
    class="group/sidebar-wrapper flex min-h-svh w-full has-[[data-variant=inset]]:bg-sidebar"
    style="--sidebar-width: 18rem; --sidebar-width-icon: 3rem;"
  >
    <ul
      class="flex w-full min-w-0 flex-col gap-1"
      data-sidebar="menu"
    >
      <li
        class="group/menu-item relative"
        data-sidebar="menu-item"
      >
        <a
          class="peer/menu-button flex w-full items-center gap-2 overflow-hidden rounded-md p-2 text-left outline-none ring-sidebar-ring transition-[width,height,padding] active:bg-sidebar-accent active:text-sidebar-accent-foreground disabled:pointer-events-none disabled:opacity-50 group-has-[[data-sidebar=menu-action]]/menu-item:pr-8 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[active=true]:bg-sidebar-accent data-[active=true]:font-medium data-[active=true]:text-sidebar-accent-foreground data-[state=open]:hover:bg-sidebar-accent data-[state=open]:hover:text-sidebar-accent-foreground group-data-[collapsible=icon]:!size-8 [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0 hover:bg-sidebar-accent hover:text-sidebar-accent-foreground h-12 text-sm group-data-[collapsible=icon]:!p-0 focus-visible:ring-primary focus-visible:ring-1"
          data-active="false"
          data-sidebar="menu-button"
          data-size="lg"
          href="foo.bar"
        >
          <svg />
          <span>
            Foo
          </span>
        </a>
      </li>
      <li
        class="group/menu-item relative"
        data-sidebar="menu-item"
      >
        <a
          class="peer/menu-button flex w-full items-center gap-2 overflow-hidden rounded-md p-2 text-left outline-none ring-sidebar-ring transition-[width,height,padding] active:bg-sidebar-accent active:text-sidebar-accent-foreground disabled:pointer-events-none disabled:opacity-50 group-has-[[data-sidebar=menu-action]]/menu-item:pr-8 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[active=true]:bg-sidebar-accent data-[active=true]:font-medium data-[active=true]:text-sidebar-accent-foreground data-[state=open]:hover:bg-sidebar-accent data-[state=open]:hover:text-sidebar-accent-foreground group-data-[collapsible=icon]:!size-8 [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0 hover:bg-sidebar-accent hover:text-sidebar-accent-foreground h-12 text-sm group-data-[collapsible=icon]:!p-0 focus-visible:ring-primary focus-visible:ring-1"
          data-active="false"
          data-sidebar="menu-button"
          data-size="lg"
          href="bar.baz"
        >
          <svg />
          <span>
            Bar
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<MenuItems /> tests > Should match the snapshot with the default args 1`] = `
<div>
  <div
    class="group/sidebar-wrapper flex min-h-svh w-full has-[[data-variant=inset]]:bg-sidebar"
    style="--sidebar-width: 18rem; --sidebar-width-icon: 3rem;"
  >
    <ul
      class="flex w-full min-w-0 flex-col gap-1"
      data-sidebar="menu"
    >
      <li
        class="group/menu-item relative"
        data-sidebar="menu-item"
      >
        <a
          class="peer/menu-button flex w-full items-center gap-2 overflow-hidden rounded-md p-2 text-left outline-none ring-sidebar-ring transition-[width,height,padding] active:bg-sidebar-accent active:text-sidebar-accent-foreground disabled:pointer-events-none disabled:opacity-50 group-has-[[data-sidebar=menu-action]]/menu-item:pr-8 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[active=true]:bg-sidebar-accent data-[active=true]:font-medium data-[active=true]:text-sidebar-accent-foreground data-[state=open]:hover:bg-sidebar-accent data-[state=open]:hover:text-sidebar-accent-foreground group-data-[collapsible=icon]:!size-8 group-data-[collapsible=icon]:!p-2 [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0 hover:bg-sidebar-accent hover:text-sidebar-accent-foreground h-8 text-sm focus-visible:ring-primary focus-visible:ring-1"
          data-active="false"
          data-sidebar="menu-button"
          data-size="default"
          href="foo.bar"
        >
          <svg />
          <span>
            Foo
          </span>
        </a>
      </li>
      <li
        class="group/menu-item relative"
        data-sidebar="menu-item"
      >
        <a
          class="peer/menu-button flex w-full items-center gap-2 overflow-hidden rounded-md p-2 text-left outline-none ring-sidebar-ring transition-[width,height,padding] active:bg-sidebar-accent active:text-sidebar-accent-foreground disabled:pointer-events-none disabled:opacity-50 group-has-[[data-sidebar=menu-action]]/menu-item:pr-8 aria-disabled:pointer-events-none aria-disabled:opacity-50 data-[active=true]:bg-sidebar-accent data-[active=true]:font-medium data-[active=true]:text-sidebar-accent-foreground data-[state=open]:hover:bg-sidebar-accent data-[state=open]:hover:text-sidebar-accent-foreground group-data-[collapsible=icon]:!size-8 group-data-[collapsible=icon]:!p-2 [&>span:last-child]:truncate [&>svg]:size-4 [&>svg]:shrink-0 hover:bg-sidebar-accent hover:text-sidebar-accent-foreground h-8 text-sm focus-visible:ring-primary focus-visible:ring-1"
          data-active="false"
          data-sidebar="menu-button"
          data-size="default"
          href="bar.baz"
        >
          <svg />
          <span>
            Bar
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
